### PR TITLE
Feature service manager v3

### PR DESCRIPTION
**Pull request by awartoft**

### CODE REVIEW

**Comment by awartoft** [comment]
@hotas2k4 just noticed i have the wrong license header

File: src/Writer/FilterPluginManager.php [hunk 1]
```diff
@@ -1,61 +1,15 @@
...
 
 namespace Zend\Log\Writer;
 
-use Zend\Log\Exception;
-use Zend\Log\Filter;
-use Zend\ServiceManager\AbstractPluginManager;
-
-class FilterPluginManager extends AbstractPluginManager
+/**
+ * @deprecated has simply be moved to the parent directory
```
Comment (weierophinney) [inline src/Writer/FilterPluginManager.php:13]: Why was it moved? Is there a compelling reason to do so?

File: src/Writer/FormatterPluginManager.php [hunk 2]
```diff
@@ -1,61 +1,15 @@
...
 
 namespace Zend\Log\Writer;
 
-use Zend\Log\Exception;
-use Zend\Log\Formatter;
-use Zend\ServiceManager\AbstractPluginManager;
-
-class FormatterPluginManager extends AbstractPluginManager
+/**
+ * @deprecated Simply moved to the parent directory
```
Comment (weierophinney) [inline src/Writer/FormatterPluginManager.php:13]: Same question here; why move/rename the plugin managers?

File: src/WriterPluginManager.php [hunk 3]
```diff
@@ -17,53 +19,47 @@
...
-     *
-     * Checks that the writer loaded is an instance of Writer\WriterInterface.
-     *
-     * @param  mixed $plugin
-     * @return void
-     * @throws Exception\InvalidArgumentException if invalid
+     * {@inheritdoc}
      */
-    public function validatePlugin($plugin)
+    public function validate($plugin)
```
Comment (weierophinney) [inline src/WriterPluginManager.php]: Use the parent implementation of this method, and define the `$instanceOf` property to detail the interface to check against.

File: src/ProcessorPluginManager.php [hunk 4]
```diff
@@ -10,48 +10,39 @@
...
-     *
-     * Checks that the processor loaded is an instance of Processor\ProcessorInterface.
-     *
-     * @param  mixed $plugin
-     * @return void
-     * @throws Exception\InvalidArgumentException if invalid
+     * {@inheritdoc}
      */
-    public function validatePlugin($plugin)
+    public function validate($plugin)
```
Comment (weierophinney) [inline src/ProcessorPluginManager.php]: As noted on another plugin manager in here, no need to override the `validate` method; just define the `$instanceOf` property to the value `Processor\ProcessorInterface::class`.

File: src/FormatterPluginManager.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        Formatter\Xml::class              => InvokableFactory::class,
+        Formatter\Db::class               => InvokableFactory::class,
+        Formatter\ErrorHandler::class     => InvokableFactory::class,
+        Formatter\ExceptionHandler::class => InvokableFactory::class,
+    ];
+
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($plugin)
```
Comment (weierophinney) [inline src/FormatterPluginManager.php]: Same comment on this one.

File: src/FilterPluginManager.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        Filter\Priority::class       => InvokableFactory::class,
+        Filter\Regex::class          => InvokableFactory::class,
+        Filter\SuppressFilter::class => InvokableFactory::class,
+        Filter\Validator::class      => InvokableFactory::class,
+    ];
+
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($plugin)
```
Comment (weierophinney) [inline src/FilterPluginManager.php]: Same comment here.

File: src/Writer/FilterPluginManager.php (hunk 1 above)
Comment (awartoft) [inline src/Writer/FilterPluginManager.php:13]: Because it's in the wrong folder, look at the folder structure and you will probably understand.

Comment (weierophinney) [inline src/Writer/FilterPluginManager.php:13]: @macnibblet That's not my point. Moving them breaks backwards compatibility, and I'd like to reduce the number of such breakages. That said, your strategy does work for that, as it's having the originals extend the new ones. So, objection withdrawn. :smile:

**Pull request merged**